### DISCUSSION TITLE
Self contained programs (vs installers).

### DISCUSSION THREAD

**Discussion by vromero** [megastep/makeself#298] [Ideas]
This is just an idea proposal to support distribution of actual programs with makeself vs containers.

I have a command line tool, multiple files, etc. That for the sake of the CI systems, I'd like to distribute as a single `.sh`. 

I've tried `makeself` and it worked great, but there are a couple of gimmicks here and there:
- the generated file has its own cli and help, this would be confusing to the use of my cli tool that of course has its own commands and help.
- the `--quiet` option cannot be passed by default and therefore the decompression and checking part is not hidden.

Perhaps this falls out of the scope of makeself, but I personally found it interesting enough to be proposed.

Thanks

**Comment by refi64** [comment]
Question: couldn't you also use AppImages for this?

**Comment by megastep** [comment]
This is an interesting concept, and one for which I'd be open to make improvements to Makeself for it. As you know it was mostly designed as a packaging format which in turns could be used to run a proper installer.

I was not familiar with the AppImages thing, but it looks interesting. Reminds me a little bit of the way apps are bundled on macOS for instance. I have shipped a lot of native binaries for various Linux/Unix distributions via Makeself, and there are usually a lot of issues with dependencies (particularly shared system libraries) that can get in the way. This could possibly work well for mostly scripted apps, or apps that have been linked statically to reduce any dependencies though.

**Comment by vromero** [comment]
> Question: couldn't you also use AppImages for this?

Yes, although that would only work in Linux.

> I was not familiar with the AppImages thing, but it looks interesting. Reminds me a little bit of the way

I was specifically thinking in this whole new world of Java apps bundled with a very small VM thanks to jlink. It seemed a low hanging fruit, specially due to the fact that there is already a maven plugin supporting makeself and hence my idea/proposal.

**Comment by mttcr** [comment]
Hello, I am adding my two cents here.

I discovered makeself today and it is a really interesting and powerful project!
My use case is similar to this discussion. My main goal is to have a single file that acts as a binary, extracts on the fly, and interprets the command line arguments. The use of makeself is almot perfect, the only drawback i have is that I have to add "--" as the first argument to "give" the remaining ones to my script. It would be ideal to have an option that allows the generated "run" file to not try to interpret args, but instead give them all to the script. Something like
at build `./makeself-2.7.1/makeself.sh **--direct-args** --nox11 --target .myProgramV3.14  --notemp --nooverwrite myProgram "My Program" ./myProgram.sh`
at run time : `./myProgram arg1 arg2`
and the expected behavior: on first run, extract program into `.myProgramV3.14`, run `./myProgram.sh arg1 arg2`, and on next launch, directly call the .sh

For the story, i wrote a minimal java app that I need to deploy as a **single file**. I am using graalvm for now, but generated output is twice as big as a makeself with a (minimal with jlink) jre and my jar, and i have compatibility issues with the generated binary (archs, glibc versions, take care of all reflection / jni calls...)
